### ISSUE TITLE
discrepancy of softmax dimension in the encoding code?

### ISSUE THREAD

**Issue by d-li14** [zhanghang1989/PyTorch-Encoding#114]
As stated by the description of Encoding equation, I assume that the softmax function is to be applied to the dimension along K.
But in this [line](https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/nn/encoding.py#L104) to calculate the assignment weights, Softmax is calculated with the parameter `dim=1` i.e. the dimension along N, noticing that the input tensor X is BxNxK size claimed by the [ScaleL2 function](https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/functions/encoding.py#L94). Would you please check is there any confusion? @zhanghang1989

**Comment by zhanghang1989** [comment]
Hi, The dim is `NxKxD`, when applying softmax with `dim=1`, which is along the `K` dimension. 
https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/nn/encoding.py#L104

**Comment by zhanghang1989** [comment]
ScaledL2 gives the pairwise distance between each input and codeword, which has the dimension of `BxKxD`

**Comment by d-li14** [comment]
@zhanghang1989 Thanks for your timely reply, but I'm sorry that I'm still confused. 
ScaledL2 computes this scalar <a href="https://www.codecogs.com/eqnedit.php?latex=sl_{ik}&space;=&space;s_k&space;\|x_i-c_k\|^2" target="_blank"><img src="https://latex.codecogs.com/gif.latex?sl_{ik}&space;=&space;s_k&space;\|x_i-c_k\|^2" title="sl_{ik} = s_k \|x_i-c_k\|^2" /></a> i.e. the internal term of softmax ![image](https://user-images.githubusercontent.com/25176159/44928250-e80b9c80-ad89-11e8-8f7b-7d184b53d981.png), <a href="https://www.codecogs.com/eqnedit.php?latex=sl_{ik}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?sl_{ik}" title="sl_{ik}" /></a>  is indeed an scalar element in a `NxK` matrix, for the subscript i ranges from 1 to N, k from 1 to K. Thus in my opinion scaledL2 will not give the size `BxKxD`. The minor misleading may be ScaledL2 gives a tensor with dimension of `BxKxN` rather than `BxNxK` as declared in the [code annotation](https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/functions/encoding.py#L94) in the condition that the inside implementation is alright.

**Comment by zhanghang1989** [comment]
Thanks @d-li14 ! This seems a bug introduced in commit https://github.com/zhanghang1989/PyTorch-Encoding/commit/8425bd86d403861f4691814f9e6038da3acbc355#diff-b25ae329176cbdc151c2c8768339698b , and it had never been detected.  Thank you very much for pointing out.

**Comment by d-li14** [comment]
OK, I have created a new pull request with respect to this problem : ) @zhanghang1989
